### PR TITLE
[FIX] payment_stripe: adapt stripe to new payment form

### PR DESCRIPTION
**Pull request by vchu-odoo**

### CODE REVIEW

File: addons/payment_stripe/views/payment_templates.xml [hunk 1]
```diff
@@ -31,4 +19,16 @@
         <script type="text/javascript" src="https://js.stripe.com/v3/"></script>
     </template>
 
+    <template id="payment_adyen.method_form" inherit_id="payment.method_form">
+        <xpath expr="//input[@name='o_payment_radio']" position="attributes">
+            <attribute name="t-att-data-stripe-publishable-key">
+                provider_sudo._stripe_get_publishable_key()
+            </attribute>
```
Comment (chevalierv) [inline addons/payment_stripe/views/payment_templates.xml]: The publishable key should already be in `_stripe_get_inline_form_values`

File: addons/payment_stripe/views/payment_templates.xml [hunk 2]
```diff
@@ -31,4 +19,16 @@
         <script type="text/javascript" src="https://js.stripe.com/v3/"></script>
     </template>
 
+    <template id="payment_adyen.method_form" inherit_id="payment.method_form">
```
Comment (chevalierv) [inline addons/payment_stripe/views/payment_templates.xml]: Maybe it's a typing error, maybe it's currently the only way to make it works ^^
```suggestion
    <template id="payment_stripe.method_form" inherit_id="payment.method_form">
```

File: addons/payment_stripe/models/payment_transaction.py [hunk 3]
```diff
@@ -143,7 +143,7 @@ def _stripe_prepare_payment_intent_payload(self):
             'currency': self.currency_id.name.lower(),
             'description': self.reference,
             'capture_method': 'manual' if self.provider_id.capture_manually else 'automatic',
-            'automatic_payment_methods[enabled]': True,
+            'expand[]': 'payment_method',
```
Comment (chevalierv) [inline addons/payment_stripe/models/payment_transaction.py:152]: why?

File: addons/payment_stripe/models/payment_transaction.py [hunk 4]
```diff
@@ -126,7 +126,7 @@ def _stripe_prepare_setup_intent_payload(self):
         return {
             'customer': customer['id'],
             'description': self.reference,
-            'automatic_payment_methods[enabled]': True,
+            # 'payment_method_types[]': self.payment_method_code,
```
Comment (chevalierv) [inline addons/payment_stripe/models/payment_transaction.py]: Does it work with this ? :eyes:

File: addons/payment_stripe/static/src/js/stripe_mixin.js [hunk 5]
```diff
@@ -2,85 +2,69 @@
...
+        const stripeInlineForm = inlineForm.querySelector('[name="o_stripe_element_container"]');
+        this.isValidation = Boolean(this.txContext['mode'] === 'validation');
+        this.stripeInlineFormValues = JSON.parse(radio.dataset['inlineFormValues']);
+        this.stripeInlineFormValues['paymentMethodTypes'] = [paymentMethodCode]
 
         // Instantiate the payment element.
-        this.stripeJS = Stripe(
-            this.stripeInlineFormValues['publishable_key'],
-            // The values required by Stripe Connect are inserted into the dataset.
-            new StripeOptions()._prepareStripeOptions(stripeInlineForm.dataset),
```
Comment (chevalierv) [inline addons/payment_stripe/static/src/js/stripe_mixin.js:46]: We need those informations for Stripe Connect, this code is overridden in our internal repo :)

File: addons/payment_stripe/static/src/js/stripe_mixin.js [hunk 6]
```diff
@@ -166,8 +164,24 @@ export default {
      *
      * @private
      * @param {object} processingValues - The processing values of the transaction.
+     * @param {number} paymentOptionId - The id of the payment option handling the transaction.
+     * @param {string} paymentMethodCode - The code of the selected payment method, if any.
      * @return {object} The processing error, if any.
      */
-    async _stripeConfirmIntent(processingValues) {},
+    async _stripeConfirmIntent(processingValues, paymentOptionId, paymentMethodCode) {
```
Comment (chevalierv) [inline addons/payment_stripe/static/src/js/stripe_mixin.js]: It looks like it's not used here (at the moment) 
```suggestion
    async _stripeConfirmIntent(processingValues, paymentOptionId,) {
```

File: addons/payment_stripe/static/src/js/stripe_mixin.js [hunk 7]
```diff
@@ -2,85 +2,69 @@
...
         paymentElement.mount(stripeInlineForm);
 
-        const tokenizationCheckbox = document.getElementById(
-            `o_payment_provider_inline_${this.formType}_form_${paymentOptionId}`
-        ).querySelector('input[name="o_payment_save_as_token"]');
-        if (this.formType === 'checkout' && tokenizationCheckbox) {
-            // Disable the tokenization checkbox for non-compatible payment methods.
-            paymentElement.addEventListener('change', ev => {
-                this.selectedPaymentMethod = ev.value.type;
-                if (this.stripeInlineFormValues['payment_methods_tokenization_support']
```
Comment (chevalierv) [inline addons/payment_stripe/static/src/js/stripe_mixin.js:71]: You can remove that from `_stripe_get_inline_form_values` also ^^

File: addons/payment_stripe/static/src/js/stripe_mixin.js [hunk 8]
```diff
@@ -2,85 +2,69 @@
...
-        this.stripeJS = Stripe(
-            this.stripeInlineFormValues['publishable_key'],
-            // The values required by Stripe Connect are inserted into the dataset.
-            new StripeOptions()._prepareStripeOptions(stripeInlineForm.dataset),
-        );
-        this.stripeElements = this.stripeJS.elements(this._getElementsOptions());
-        this.stripeElements.providerId = paymentOptionId;
+        this.stripeJS = Stripe(radio.dataset['stripePublishableKey']);
+        this.stripePaymentElements[paymentOptionId] = this.stripeJS.elements(this._getElementsOptions());
+        this.stripePaymentElements[paymentOptionId].providerId = providerId;
```
Comment (chevalierv) [inline addons/payment_stripe/static/src/js/stripe_mixin.js]: Is this really still needed?

**Review by chevalierv — COMMENTED** [review]
Hey @vchu-odoo 

I couldn't refrain from having a look at this PR :sweat_smile: 

Thanks for your work on this, I like the way it's going with the new implementation! 

I've left a few comments to be resolved before or when you'll merge this onto ANV's branch.

Have a nice day :sun_with_face:

File: addons/payment_stripe/models/payment_transaction.py (hunk 3 above)
Comment (vchu-odoo) [reply addons/payment_stripe/models/payment_transaction.py:152]: to get payment method in expanded format to retrieve the payment method type in line 360

Comment (chevalierv) [reply addons/payment_stripe/models/payment_transaction.py:152]: I thought it was done in the controller :thinking:  It's great if it works like this ^^

Comment (AntoineVDV) [reply addons/payment_stripe/models/payment_transaction.py:152]: It needs to be done here too when paying by token.

**Review by AntoineVDV — APPROVED** [review]

**Pull request merged**